04-cs-fundamentals/internship/hash-tables.md

```diff
@@ -0,0 +1,319 @@
+
+Hash tables are execellent for looking up data from a key.  Binary search trees are excellent data structures for maintaining items in order while linked lists provide quick access to the front and rear and maintain a specific order.
+
+So, which would you use in these situations?
+
+**Question 1:**
+
+<details>
+  <summary>A list of candidate records that need to be stored so that you can find the max or min test scores</summary>
+  Hash tables are not good for maintaining ordered data.  Instead a binary search tree, or sorted array or linked list (sorted by test scores) would serve better.
+</details>
+
+<!-- source: http://www.cs.cmu.edu/~guna/15-123S11/Lectures/Lecture17.pdf -->
+<br />
+
+**Question 2:**
+
+<details>
+  <summary>Fedex needs to provide quick access for customers to check the status of their packages.  Quick lookup time using distinct tracking numbers is important.</summary>
+  Since the tracking numbers are distinct and order is not important, a hash table makes the most sense because of it's quick lookup time.
+</details>
+
+<br />
+
+**Question 3:**
+
+<details>
+  <summary>Postgres needs to keep track of fields in a table to output sorted data.  Items will be regularly inserted and deleted from the table.</summary>
+  Since order is important, maintaining a set of binary search trees would make a great deal of sense.
+</details>
+
+## Summary
+
+Hashes are one of the most practically useful data structures you will encounter.  They make a great many problems easier and with lower time complexity.  You will also likely **never** have to create a hash table data structure yourself, but you will need to know their practical applications and be able to discuss how they function.
+
+There are many ways to structure a hash table.  One of the most common is simple chaining with each element of the internal array or _bucket_ referencing the beginning of a linked list.  Alternative solutions include _dynamic arrays_ which use arrays to store the elements in each bucket and open addressing using either _linear probling_, _quadratic probing_ or _double hashing_.  It's also important to recognize that a hash table depends on **unique** keys.  If the keys for items are not unique there is no way to generate a unique hash result.
+
+
+<!-- Image from: https://www.bigocheatsheet.com/ -->
+
+## Terms & Definitions
+
+| Term | Definition |
+|-- |-- |
+| **Bucket** | An element in the internal array used by a hash table. |
+| **Chaining** | A strategy to handle collisions where each element of a hash table's internal array serves as the head of a linked list.  This allows the same bucket to store multiple items.
+| **Clustering** | When multiple elements are mapped to the same bucket in a hash table |
+| **Collision** | When two keys are mapped to the same bucket in a hash table. |
+| **Double Hashing** | A strategy to handle collisions in a hash table where a secondary hash function is used when a collision occurs. |
```

Comment:
I'd like to see "Chaining", "Double Hashing", "Linear Probing", in adjacent rows in the table (instead of alphabetical order).